### PR TITLE
Behavior annex plugin: improve syntax highlighting and error reporting

### PR DESCRIPTION
**Pull request by Etienne13**

### CODE REVIEW

**Comment by Etienne13** [comment]
Behavior annex: 
 - enable syntax highlighting in case of syntax or name resolution errors
 - prevent the parser from reporting the entire annex block as erroneous on syntax errors

**Comment by Etienne13** [comment]
@lwrage not sure why the last commit cannot be built. Is it because the static method convertToIResource (class OsateResourceUtil) is deprecated? Should I add an annotation or do you see a better way to do this?

**Comment by Etienne13** [comment]
I found the recommendation to use toIFile instead of  convertToIResource in the comments. So I did the replacement.

**Comment by Etienne13** [comment]
I reverted the last two commit because of the comments in issue #891

**Comment by Etienne13** [comment]
@lwrage can you update me on the status of this pull request?

**Comment by lwrage** [comment]
@Etienne13 We've merged #1905 to propagate annex errors and warnings. Could you update this PR and test that it still works, please?

**Comment by Etienne13** [comment]
@lwrage Just did a merge of the master branch of osate2 in my fork. I had to fix a conflict and adapt my implementation. The PR is updated and tested: (i) syntax highlighting is visible in case of syntax/resolution error and (ii) the annexSection is set to null in case of error (thus removing the meaningless error message saying the initialState attribute cannot be null). However the checks performed by Jenkins failed; don't know why. Can you help me with that?

File: ba/org.osate.ba/src/org/osate/ba/texteditor/XtextAadlBaHighlighter.java [hunk 1]
```diff
@@ -58,4 +60,24 @@ public void addToHighlighting(int annexOffset, int relativeOffset, int length, S
...
+
+  private static Map<AnnexSubclause, XtextAadlBaHighlighter> _highlighterPerAnnex = 
+      new HashMap<AnnexSubclause, XtextAadlBaHighlighter>();
+  
+  public static XtextAadlBaHighlighter getHighlighter(AnnexSubclause as)
+  {
+    if(_highlighterPerAnnex.get(as)==null)
+    {
+      XtextAadlBaHighlighter ht = new XtextAadlBaHighlighter();
+      _highlighterPerAnnex.put(as, ht);
```
Comment (lwrage) [inline ba/org.osate.ba/src/org/osate/ba/texteditor/XtextAadlBaHighlighter.java]: Why are highlighters stored in a has map? Each run of the parser creates new AnnexSubclause objects, so new highlighters will be created for each parser run.

**Comment by lwrage** [comment]
There is a test failure. Project org.osate.ba.tests, file Issue1884Test.xtend. The test needs to be adapted to the changes in this PR.

File: ba/org.osate.ba/src/org/osate/ba/texteditor/XtextAadlBaHighlighter.java (hunk 1 above)
Comment (Etienne13) [reply ba/org.osate.ba/src/org/osate/ba/texteditor/XtextAadlBaHighlighter.java]: The reason for using a map is the following: text to highlight in the BA subclause is identified during parsing (when the parser is executed from AnnexParserAgent class) but syntax highlighting is processed latter on (when highlightAnnexSubclause is called in class AnnexHighlighterProxy). So we need to save the locations to highlight during parsing and retreive them during highlighting.

The only shared (and stable) information between these methods invocation is the AnnexSubclause (or AnnexLibrary) object. This is why we store the locations to highlight in a map where the AnnexSubclause object is the key. This is also the reason why we store the current AnnexSubclause object in a global variable before starting the parsing. In AnnexParserAgent:

`if(defaultAnnexSection instanceof AnnexSubclause)
					AnnexUtil.setCurrentAnnexSubclause((AnnexSubclause) defaultAnnexSection);
`

I agree it is not very clean but given the methods provided by OSATE for syntax hihglighting of ANTLR based annexes I did not find a better solution so far.

File: ba/org.osate.ba/src/org/osate/ba/texteditor/XtextAadlBaHighlighter.java [hunk 2]
```diff
@@ -58,4 +60,24 @@ public void addToHighlighting(int annexOffset, int relativeOffset, int length, S
               id));
   }
   
+  private XtextAadlBaHighlighter() {}
+
+  private static Map<AnnexSubclause, XtextAadlBaHighlighter> _highlighterPerAnnex = 
+      new HashMap<AnnexSubclause, XtextAadlBaHighlighter>();
```
[the remark this answers is not in the document]
Comment (Etienne13) [reply ba/org.osate.ba/src/org/osate/ba/texteditor/XtextAadlBaHighlighter.java]: I agree with the comment and the proposal, changed it in commit 2d1e528

**Comment by Etienne13** [comment]
> There is a test failure. Project org.osate.ba.tests, file Issue1884Test.xtend. The test needs to be adapted to the changes in this PR.

I understand, but I do not know how to run this test. Can you explain me? Or do you want to update the test?

**Comment by lwrage** [comment]
You can run all tests in a test plugin: right click on the project, select run as -> junit plug-in test
You can also run a single test when you right click on the xtend file. Tests are written in xtend but can also be written in plain Java (xtend is precompiled to Java).

**Comment by Etienne13** [comment]
@lwrage thanks, the test is passing now.

**Pull request merged**